### PR TITLE
feat: [IDP-1274]: IDP Licensing - License usage

### PR DESCRIPTION
**Pull request by sathish-soundarapandian**
## Describe your changes
IDP Licensing - License usage

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/54099)

### CODE REVIEW

**Review by vikyathharekal — COMMENTED** [review]
> Reviewed 38 of 43 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/54099)* status: Some of the checks are too old. 43 file(s) to review. 0 of 2 approvals obtained. 

___
*[`idp-service/src/main/java/io/harness/idp/events/consumers/AbstractIdpServiceMiscRedisStreamConsumer.java` line 40 at r1](https://reviewable.io/reviews/harness/harness-core/54099#-Ngqc65L-bZdW5ZU7osQ:-Ngqc65L-bZdW5ZU7osR:b-iwooz8) ([raw file](https://github.com/harness/harness-core/blob/e6460cec10a5a6663e393cfd4468fc81643bd3ce/idp-service/src/main/java/io/harness/idp/events/consumers/AbstractIdpServiceMiscRedisStreamConsumer.java#L40)):*
> ```Java
> @Slf4j
> @OwnedBy(HarnessTeam.IDP)
> public abstract class AbstractIdpServiceMiscRedisStreamConsumer extends RedisTraceConsumer implements IdpRedisConsumer {
> ```

Do we see any other use case where we might extend this? Or is licence usage the only one?
Or can our existing Consumers extend this?



_Code quote:_
```Java
AbstractIdpServiceMiscRedisStreamConsumer
```
___
*[`idp-service/src/main/java/io/harness/idp/license/usage/entities/ActiveDevelopersDailyCountEntity.java` line 49 at r1](https://reviewable.io/reviews/harness/harness-core/54099#-NgqgWy5A8eGplZ3io4I:-NgqgWy5A8eGplZ3io4J:byh58za) ([raw file](https://github.com/harness/harness-core/blob/e6460cec10a5a6663e393cfd4468fc81643bd3ce/idp-service/src/main/java/io/harness/idp/license/usage/entities/ActiveDevelopersDailyCountEntity.java#L49)):*
> ```Java
> @HarnessEntity(exportable = true)
> @OwnedBy(HarnessTeam.IDP)
> public class ActiveDevelopersDailyCountEntity implements PersistentEntity, CreatedAtAware {
> ```

What about TTL? Would it be 6 or 7 months?
We could set like this
```@Builder.Default @FdTtlIndex Date validUntil = Date.from(OffsetDateTime.now().plusMonths(7).toInstant());```



_Code quote:_
```Java
ActiveDevelopersDailyCountEntity
```
___
*[`idp-service/src/main/java/io/harness/idp/license/usage/entities/ActiveDevelopersEntity.java` line 51 at r1](https://reviewable.io/reviews/harness/harness-core/54099#-Ngqh2uO3MSt2MFmVVMW:-Ngqh2uO3MSt2MFmVVMX:bx6xdqb) ([raw file](https://github.com/harness/harness-core/blob/e6460cec10a5a6663e393cfd4468fc81643bd3ce/idp-service/src/main/java/io/harness/idp/license/usage/entities/ActiveDevelopersEntity.java#L51)):*
> ```Java
> @HarnessEntity(exportable = true)
> @OwnedBy(HarnessTeam.IDP)
> public class ActiveDevelopersEntity implements PersistentEntity, CreatedAtAware, UpdatedAtAware {
> ```

Do we need to retain this data after rolling it up to a day?
What if the user no longer is part of the org? Do we plan to clean up from our collection?



_Code quote:_
```Java
ActiveDevelopersEntity
```
___
*[`idp-service/src/main/java/io/harness/idp/license/usage/jobs/LicenseUsageDailyCountJob.java` line 45 at r1](https://reviewable.io/reviews/harness/harness-core/54099#-NgqoWsn9c2kU2bCDNt2:-NgqoWsn9c2kU2bCDNt3:bpibzb4) ([raw file](https://github.com/harness/harness-core/blob/e6460cec10a5a6663e393cfd4468fc81643bd3ce/idp-service/src/main/java/io/harness/idp/license/usage/jobs/LicenseUsageDailyCountJob.java#L45)):*
> ```Java
>     executorService = Executors.newSingleThreadScheduledExecutor(
>         new ThreadFactoryBuilder().setNameFormat("license-usage-daily-count-job").build());
>     long midnight = LocalDateTime.now().until(LocalDate.now().plusDays(1).atStartOfDay(), ChronoUnit.MINUTES);
> ```

If we deploy at 6pm UTC. This job will be run after 6 hours at midnight. Is that correct?
___
*[`idp-service/src/main/java/io/harness/idp/license/usage/resources/IDPLicenseUsageResource.java` line 57 at r1](https://reviewable.io/reviews/harness/harness-core/54099#-NgqqAraEbANoJxekNHC:-NgqqAraEbANoJxekNHD:b-wac79y) ([raw file](https://github.com/harness/harness-core/blob/e6460cec10a5a6663e393cfd4468fc81643bd3ce/idp-service/src/main/java/io/harness/idp/license/usage/resources/IDPLicenseUsageResource.java#L57)):*
> ```Java
> import org.springframework.data.domain.Sort;
> 
> @Api("/usage")
> ```

Will these APIs be used in Harness License page? Can't these be spec first?



_Code quote:_
```Java
@Api("/usage")
```

**Review by sathish-soundarapandian — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/54099)* status: Some of the checks are too old. 43 file(s) to review. 5. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`idp-service/src/main/java/io/harness/idp/events/consumers/AbstractIdpServiceMiscRedisStreamConsumer.java` line 40 at r1](https://reviewable.io/reviews/harness/harness-core/54099#-Ngqc65L-bZdW5ZU7osQ:-Ngr-cHA6zpczaSszXwf:b-43s8w7) ([raw file](https://github.com/harness/harness-core/blob/e6460cec10a5a6663e393cfd4468fc81643bd3ce/idp-service/src/main/java/io/harness/idp/events/consumers/AbstractIdpServiceMiscRedisStreamConsumer.java#L40)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

Do we see any other use case where we might extend this? Or is licence usage the only one?
Or can our existing Consumers extend this?
</blockquote></details>

Two reasons why I did this way - 

1. Our EntityCrudStreamConsumer is solely based on entity_crud topic. With this license usage eventing part, we don't need to put the data to entity_crud as it is not exactly an entity crud. It's just an normal event data and it can belong to it's own topic definition. I see there is a mistake in asyncCatalogImport. I have done it as entity_crud, but eventually that should move to it's own topic definition

2. Yes. We can extend this for other use-case. Our current consumers  - userGroup, connector, secret are entity_crud and asyncCatalogImport, licenseUsageEvent are idp-service misc use-cases. This switch cases for handler and event-handler pattern is Ok, but from design perspective and readability - we can move to this AbstractIdpServiceMiscRedisStreamConsumer pattern (may be name it as AbstractIdpServiceRedisStreamConsumer) and entity_crud, other misc event topics can also move to this pattern.

Let me know about your thoughts on this
___
*[`idp-service/src/main/java/io/harness/idp/license/usage/entities/ActiveDevelopersDailyCountEntity.java` line 49 at r1](https://reviewable.io/reviews/harness/harness-core/54099#-NgqgWy5A8eGplZ3io4I:-NgqzP1d03LDATWNdTCw:bntwvtz) ([raw file](https://github.com/harness/harness-core/blob/e6460cec10a5a6663e393cfd4468fc81643bd3ce/idp-service/src/main/java/io/harness/idp/license/usage/entities/ActiveDevelopersDailyCountEntity.java#L49)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

What about TTL? Would it be 6 or 7 months?
We could set like this
```@Builder.Default @FdTtlIndex Date validUntil = Date.from(OffsetDateTime.now().plusMonths(7).toInstant());```
</blockquote></details>

Done. Have set it as 6 months for now.

Some numbers - let's say we have 200 accounts. At max, each account will have 180 rows and that brings the total to 36000 rows in this collection. We should be good with 6 months. Anyways, we publish these data to reporting sources and there the cumulative period can be even more as there the comparison will be even on yearly data.
___
*[`idp-service/src/main/java/io/harness/idp/license/usage/entities/ActiveDevelopersEntity.java` line 51 at r1](https://reviewable.io/reviews/harness/harness-core/54099#-Ngqh2uO3MSt2MFmVVMW:-Ngr9p452kcrmi0q0_kU:b-azaxvu) ([raw file](https://github.com/harness/harness-core/blob/e6460cec10a5a6663e393cfd4468fc81643bd3ce/idp-service/src/main/java/io/harness/idp/license/usage/entities/ActiveDevelopersEntity.java#L51)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

Do we need to retain this data after rolling it up to a day?
What if the user no longer is part of the org? Do we plan to clean up from our collection?
</blockquote></details>

As of now, there is no clean up plan in this collection. Again going with average numbers, 100 accounts with 1000 users on each account. We will have close to 30 MB of data in this collection (each record of 300 bytes)

Even if user is no longer part of the org, we will not cleanup that user data from ActiveDevelopers collection. Just that it will get listed in the active developer trend with lastAccessedAt in subscriptions page.

As this is ActiveDevelopers at any given point of time, we don't retain it for the day. Instead it overrides the last activity and the roll up will be stored in the daily trend collection.
___
*[`idp-service/src/main/java/io/harness/idp/license/usage/jobs/LicenseUsageDailyCountJob.java` line 45 at r1](https://reviewable.io/reviews/harness/harness-core/54099#-NgqoWsn9c2kU2bCDNt2:-NgrCGfjCUrf_xMMNkpZ:bu1bcta) ([raw file](https://github.com/harness/harness-core/blob/e6460cec10a5a6663e393cfd4468fc81643bd3ce/idp-service/src/main/java/io/harness/idp/license/usage/jobs/LicenseUsageDailyCountJob.java#L45)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

If we deploy at 6pm UTC. This job will be run after 6 hours at midnight. Is that correct?
</blockquote></details>

As we haven't specified the zoneId, the local date time will be based on the server's time. So whatever the time we deploy, it will be always on the next day 12.00 am and the job will run for previous day data. And post that, it will run each day 12.00 am.
And on the next deployment, the same pattern will follow.
Have tested it once. Will need to test it thoroughly.
___
*[`idp-service/src/main/java/io/harness/idp/license/usage/resources/IDPLicenseUsageResource.java` line 57 at r1](https://reviewable.io/reviews/harness/harness-core/54099#-NgqqAraEbANoJxekNHC:-Ngqy1OXBn9hE6pGixFf:b6d3w1v) ([raw file](https://github.com/harness/harness-core/blob/e6460cec10a5a6663e393cfd4468fc81643bd3ce/idp-service/src/main/java/io/harness/idp/license/usage/resources/IDPLicenseUsageResource.java#L57)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

Will these APIs be used in Harness License page? Can't these be spec first?
</blockquote></details>

Yes. These APIs will be used in Harness License page. More specifically in the subscriptions page where we will have IDP and on click of that, these APIs will be called.

Initially I thought about spec first approach, but the problem here is that - these APIs request and response bodies are extended from common framework DTOs. That's why I have put it as normal API resource itself. Currently other modules also follow the same.
Again, providing this at the framework level might be complicated as well.

**Comment by sathish-soundarapandian** [comment]
> Done. Have set it as 6 months for now.

Will be set to 12 months as per discussion with Product

**Review by vikyathharekal — COMMENTED** [review]
> Reviewed 5 of 43 files at r1, 1 of 1 files at r2, 20 of 20 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/54099)* status: 45 file(s) to review. 1. unresolved discussions. 0 of 2 approvals obtained.

**Review by vikyathharekal — DISMISSED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/54099)* status: 45 file(s) to review. 1 of 2 approvals obtained.

**Comment by sathish-soundarapandian** [comment]
trigger uts

File: idp-service/contracts/openapi/v1/openapi.yaml [hunk 1]
```diff
@@ -3237,6 +3256,48 @@ components:
...
+      properties:
+        userIdentifier:
+          type: string
+          x-stoplight:
+            id: kwshwrjsp0kw0
+        email:
+          type: string
+          x-stoplight:
+            id: gg9ozxyv4slx8
+        userName:
```
Comment (mankrit-singh) [inline idp-service/contracts/openapi/v1/openapi.yaml:3274]: please change the casing here to user_name, accesed_at etc.

Comment (sathish-soundarapandian) [reply idp-service/contracts/openapi/v1/openapi.yaml]: Done

**Review by mankrit-singh — APPROVED** [review]

**Review by kapil-harness — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/54099)* status: 46 file(s) to review. 1 change requests. 1. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`953-events-api/src/main/proto/io/harness/eventsframework/schemas/idp/idp_license_usage_capture_event.proto` line 14 at r5](https://reviewable.io/reviews/harness/harness-core/54099#-Nh1IsKx43ZfgT2wesqj:-Nh1IsKx43ZfgT2wesqk:bawu5v4) ([raw file](https://github.com/harness/harness-core/blob/b6b318cedccae2e0843e969180576616757013e9/953-events-api/src/main/proto/io/harness/eventsframework/schemas/idp/idp_license_usage_capture_event.proto#L14)):*
> ```proto
> 
> message IdpLicenseUsageCaptureEvent {
>   string account_identifier = 1;
> ```

lets not use underscore, lets use accountIdentifier
same for other variables as well
___
*[`idp-service/src/main/java/io/harness/idp/app/IdpServiceRequestInterceptor.java` line 47 at r5](https://reviewable.io/reviews/harness/harness-core/54099#-Nh1JnSa65Ff-RQsJK4s:-Nh1JnSa65Ff-RQsJK4t:b-gomzsl) ([raw file](https://github.com/harness/harness-core/blob/b6b318cedccae2e0843e969180576616757013e9/idp-service/src/main/java/io/harness/idp/app/IdpServiceRequestInterceptor.java#L47)):*
> ```Java
>       }
>     } catch (Exception ex) {
>       log.error(
> ```

why not throw here?

**Review by kapil-harness — CHANGES_REQUESTED** [review]
@sathish-soundarapandian why is this PR 1700+ LOC? Why are we trying to squeeze all the changes all at once? 
For ex why are we adding stream producers/consumers with license usage, it can definitely be separated out.
Its very difficult for the reviewers to review such PRs, increases the possibility of errors.

**Comment by sathish-soundarapandian** [comment]
> @sathish-soundarapandian why is this PR 1700+ LOC? Why are we trying to squeeze all the changes all at once? For ex why are we adding stream producers/consumers with license usage, it can definitely be separated out. Its very difficult for the reviewers to review such PRs, increases the possibility of errors.

Yes. Agreed. That's too many files, line changes in one PR. 
"adding stream producers/consumers with license usage" - our architecture demanded it and that's the reason why we did this way. Our license usage needs to be derived with user interaction and we capture that in stream first and process it later on the consumer side. And on top of that, I did the changes for subscriptions page powering.

Yeah, this could have separated out. Will take care of it.

**Review by sathish-soundarapandian — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/54099)* status: 46 file(s) to review. 1 change requests. 3. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`953-events-api/src/main/proto/io/harness/eventsframework/schemas/idp/idp_license_usage_capture_event.proto` line 14 at r5](https://reviewable.io/reviews/harness/harness-core/54099#-Nh1IsKx43ZfgT2wesqj:-Nh1ag06E2t6YqS-1tID:b-wo33t3) ([raw file](https://github.com/harness/harness-core/blob/b6b318cedccae2e0843e969180576616757013e9/953-events-api/src/main/proto/io/harness/eventsframework/schemas/idp/idp_license_usage_capture_event.proto#L14)):*
<details><summary><i>Previously, sathish-soundarapandian (Sathish) wrote…</i></summary><blockquote>

As per doc, it is recommended to use snake_case for proto field name, file name
And we do use this convention.
</blockquote></details>

https://protobuf.dev/programming-guides/style/

**Comment by harness-jc** [comment]
Helm changes approved.

**Review by harness-jc — DISMISSED** [review]

**Comment by sathish-soundarapandian** [comment]
trigger uts

**Review by vikyathharekal — DISMISSED** [review]
> Reviewed 1 of 1 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/54099)* status: 46 file(s) to review. 1 change requests. 3. unresolved discussions.

**Comment by sathish-soundarapandian** [comment]
trigger review

**Comment by sathish-soundarapandian** [comment]
trigger ut0, ut1, ut2, ut3, ut4, ut5, ut6, ut7, ut8, ut9

**Review by kapil-harness — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/54099)* status: 46 file(s) to review. 1 change requests. 1. unresolved discussions. 

___
*[`953-events-api/src/main/proto/io/harness/eventsframework/schemas/idp/idp_license_usage_capture_event.proto` line 14 at r5](https://reviewable.io/reviews/harness/harness-core/54099#-Nh1IsKx43ZfgT2wesqj:-Nh1aemS0tWakfmGiWoF:b-4eoxxg) ([raw file](https://github.com/harness/harness-core/blob/b6b318cedccae2e0843e969180576616757013e9/953-events-api/src/main/proto/io/harness/eventsframework/schemas/idp/idp_license_usage_capture_event.proto#L14)):*
<details><summary><i>Previously, sathish-soundarapandian (Sathish) wrote…</i></summary><blockquote>

https://protobuf.dev/programming-guides/style/
</blockquote></details>

ok, but I have not seen even 1 instance of that in our codebase
___
*[`idp-service/src/main/java/io/harness/idp/events/producers/IdpServiceMiscRedisProducer.java` line 41 at r5](https://reviewable.io/reviews/harness/harness-core/54099#-Nh1isIREXUuxQAqZq4T:-Nh1isIS8VTKQ7GA4QxJ:b-z4lm46) ([raw file](https://github.com/harness/harness-core/blob/b6b318cedccae2e0843e969180576616757013e9/idp-service/src/main/java/io/harness/idp/events/producers/IdpServiceMiscRedisProducer.java#L41)):*
> ```Java
>       String eventId = eventProducer.send(
>           Message.newBuilder()
>               .putAllMetadata(Map.of("accountId", accountIdentifier, EventsFrameworkMetadataConstants.ENTITY_TYPE,
> ```

should be fully qualified name i.e. accountIdentifier
___
*[`idp-service/src/main/java/io/harness/idp/license/usage/jobs/LicenseUsageDailyCountJob.java` line 62 at r5](https://reviewable.io/reviews/harness/harness-core/54099#-Nh1jfi3D6m24oSYmbfM:-Nh1jfi3D6m24oSYmbfN:b-wfvn8u) ([raw file](https://github.com/harness/harness-core/blob/b6b318cedccae2e0843e969180576616757013e9/idp-service/src/main/java/io/harness/idp/license/usage/jobs/LicenseUsageDailyCountJob.java#L62)):*
> ```Java
>       idpModuleLicenseUsage.licenseUsageDailyCountAggregationPerAccount();
>     } catch (Exception ex) {
>       log.error("Error in LicenseUsageDailyCountJob. Error = {}", ex.getMessage(), ex);
> ```

we should throw here, as job which runs once daily is failing
___
*[`idp-service/src/main/java/io/harness/idp/license/usage/resources/IDPLicenseUsageResource.java` line 57 at r5](https://reviewable.io/reviews/harness/harness-core/54099#-Nh1kQOY5jR0Pon1TITJ:-Nh1kQOZB6pI-fAgeI9s:b-kawdzm) ([raw file](https://github.com/harness/harness-core/blob/b6b318cedccae2e0843e969180576616757013e9/idp-service/src/main/java/io/harness/idp/license/usage/resources/IDPLicenseUsageResource.java#L57)):*
> ```Java
> import org.springframework.data.domain.Sort;
> 
> @Api("/usage")
> ```

why are we not using spec first API approach?
___
*[`idp-service/src/main/java/io/harness/idp/license/usage/resources/IDPLicenseUsageResource.java` line 67 at r5](https://reviewable.io/reviews/harness/harness-core/54099#-Nh1kdQrBjOIuc9Tudi5:-Nh1kdQrBjOIuc9Tudi6:b-ze5h4a) ([raw file](https://github.com/harness/harness-core/blob/b6b318cedccae2e0843e969180576616757013e9/idp-service/src/main/java/io/harness/idp/license/usage/resources/IDPLicenseUsageResource.java#L67)):*
> ```Java
>   @Inject IDPModuleLicenseUsage idpModuleLicenseUsage;
> 
>   @POST
> ```

why is a list API a POST? should be GET
___
*[`idp-service/src/main/java/io/harness/idp/license/usage/resources/IDPLicenseUsageResource.java` line 94 at r5](https://reviewable.io/reviews/harness/harness-core/54099#-Nh1kij31q7FD_fm1mwA:-Nh1kij31q7FD_fm1mwB:bnj7hl0) ([raw file](https://github.com/harness/harness-core/blob/b6b318cedccae2e0843e969180576616757013e9/idp-service/src/main/java/io/harness/idp/license/usage/resources/IDPLicenseUsageResource.java#L94)):*
> ```Java
>   }
> 
>   @POST
> ```

same as above
___
*[`idp-service/src/main/java/io/harness/idp/license/usage/service/impl/IDPLicenseUsageImpl.java` line 61 at r5](https://reviewable.io/reviews/harness/harness-core/54099#-Nh1lf4cC3hDwolsZyBz:-Nh1lf4cC3hDwolsZyC-:bwo4foj) ([raw file](https://github.com/harness/harness-core/blob/b6b318cedccae2e0843e969180576616757013e9/idp-service/src/main/java/io/harness/idp/license/usage/service/impl/IDPLicenseUsageImpl.java#L61)):*
> ```Java
> 
>   @Override
>   public Page<IDPActiveDevelopersDTO> listLicenseUsage(
> ```

nit: indentation can be improved, this is very unreadable
___
*[`idp-service/src/main/java/io/harness/idp/events/consumers/AbstractIdpServiceRedisStreamConsumer.java` line 60 at r5](https://reviewable.io/reviews/harness/harness-core/54099#-Nh1h7nrErziBCvwuM4L:-Nh1h7ns3MO5fYbpTdLA:bvygk50) ([raw file](https://github.com/harness/harness-core/blob/b6b318cedccae2e0843e969180576616757013e9/idp-service/src/main/java/io/harness/idp/events/consumers/AbstractIdpServiceRedisStreamConsumer.java#L60)):*
> ```Java
>     String entityType = metadataMap.get(ENTITY_TYPE);
>     String action = metadataMap.get(ACTION);
>     if ((entityType == null || !entityType.equals(entityTypeToCheckAgainst))
> ```

.equals can handle null case, so no need for a separate null check
___
*[`idp-service/src/main/java/io/harness/idp/events/consumers/AbstractIdpServiceRedisStreamConsumer.java` line 99 at r5](https://reviewable.io/reviews/harness/harness-core/54099#-Nh1gj2zEX7Nm1mEQ8mE:-Nh1gj2zEX7Nm1mEQ8mF:bv5fsqb) ([raw file](https://github.com/harness/harness-core/blob/b6b318cedccae2e0843e969180576616757013e9/idp-service/src/main/java/io/harness/idp/events/consumers/AbstractIdpServiceRedisStreamConsumer.java#L99)):*
> ```Java
>       }
>     } catch (InterruptedException ex) {
>       SecurityContextBuilder.unsetCompleteContext();
> ```

this is already in finally clause, can be removed from here

**Review by sathish-soundarapandian — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/54099)* status: 46 file(s) to review. 1 change requests. 8. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`953-events-api/src/main/proto/io/harness/eventsframework/schemas/idp/idp_license_usage_capture_event.proto` line 14 at r5](https://reviewable.io/reviews/harness/harness-core/54099#-Nh1IsKx43ZfgT2wesqj:-Nh5Fhuw6hIXkqwzxCv5:b-7jkolf) ([raw file](https://github.com/harness/harness-core/blob/b6b318cedccae2e0843e969180576616757013e9/953-events-api/src/main/proto/io/harness/eventsframework/schemas/idp/idp_license_usage_capture_event.proto#L14)):*
<details><summary><i>Previously, kapil-harness (Kapil Choudhary) wrote…</i></summary><blockquote>

ok, but I have not seen even 1 instance of that in our codebase
</blockquote></details>

Some instances -
https://github.com/harness/harness-core/tree/develop/953-git-sync-commons/src/main/proto/io/harness/gitsync
https://github.com/harness/harness-core/tree/develop/950-delegate-tasks-beans/src/main/proto/io/harness/perpetualtask

You can search for "account_id" in *.proto from IntelliJ to some more instances.
___
*[`idp-service/src/main/java/io/harness/idp/events/producers/IdpServiceMiscRedisProducer.java` line 41 at r5](https://reviewable.io/reviews/harness/harness-core/54099#-Nh1isIREXUuxQAqZq4T:-Nh5Fppi82hp6yoyWWoT:bb6un1j) ([raw file](https://github.com/harness/harness-core/blob/b6b318cedccae2e0843e969180576616757013e9/idp-service/src/main/java/io/harness/idp/events/producers/IdpServiceMiscRedisProducer.java#L41)):*
<details><summary><i>Previously, kapil-harness (Kapil Choudhary) wrote…</i></summary><blockquote>

should be fully qualified name i.e. accountIdentifier
</blockquote></details>

Done
___
*[`idp-service/src/main/java/io/harness/idp/license/usage/jobs/LicenseUsageDailyCountJob.java` line 62 at r5](https://reviewable.io/reviews/harness/harness-core/54099#-Nh1jfi3D6m24oSYmbfM:-Nh5FsL88DG6t8XH0iTq:bb6un1j) ([raw file](https://github.com/harness/harness-core/blob/b6b318cedccae2e0843e969180576616757013e9/idp-service/src/main/java/io/harness/idp/license/usage/jobs/LicenseUsageDailyCountJob.java#L62)):*
<details><summary><i>Previously, kapil-harness (Kapil Choudhary) wrote…</i></summary><blockquote>

we should throw here, as job which runs once daily is failing
</blockquote></details>

Done
___
*[`idp-service/src/main/java/io/harness/idp/license/usage/resources/IDPLicenseUsageResource.java` line 57 at r5](https://reviewable.io/reviews/harness/harness-core/54099#-Nh1kQOY5jR0Pon1TITJ:-Nh5Fu6X4blZevthIgf3:by97xcm) ([raw file](https://github.com/harness/harness-core/blob/b6b318cedccae2e0843e969180576616757013e9/idp-service/src/main/java/io/harness/idp/license/usage/resources/IDPLicenseUsageResource.java#L57)):*
<details><summary><i>Previously, kapil-harness (Kapil Choudhary) wrote…</i></summary><blockquote>

why are we not using spec first API approach?
</blockquote></details>

Initially I thought about spec first approach, but the problem here is that - these APIs request and response bodies are extended from common framework DTOs. That's why I have put it as normal API resource itself. Will end up replicating it.

My thought here is that framework should provide some level of support here. Like "LicenseUsageResource.java" should be in open-api format and it should set standards for modules to implement their own
___
*[`idp-service/src/main/java/io/harness/idp/license/usage/resources/IDPLicenseUsageResource.java` line 67 at r5](https://reviewable.io/reviews/harness/harness-core/54099#-Nh1kdQrBjOIuc9Tudi5:-Nh5G-HT-ihMgclgQ99D:by1cq16) ([raw file](https://github.com/harness/harness-core/blob/b6b318cedccae2e0843e969180576616757013e9/idp-service/src/main/java/io/harness/idp/license/usage/resources/IDPLicenseUsageResource.java#L67)):*
<details><summary><i>Previously, kapil-harness (Kapil Choudhary) wrote…</i></summary><blockquote>

why is a list API a POST? should be GET
</blockquote></details>

We accept filter / other params in these APIs. And those are send in request body and that's why it is POST. 
As per standards, we shouldn't use request body in GET request if that has a meaning on response. 
And also as per spec standards, it is not advisable to have to many parms in query / path. Instead those can converted to request body.
___
*[`idp-service/src/main/java/io/harness/idp/license/usage/resources/IDPLicenseUsageResource.java` line 94 at r5](https://reviewable.io/reviews/harness/harness-core/54099#-Nh1kij31q7FD_fm1mwA:-Nh5G-jQ0lppgr-rNW8k:by1cq16) ([raw file](https://github.com/harness/harness-core/blob/b6b318cedccae2e0843e969180576616757013e9/idp-service/src/main/java/io/harness/idp/license/usage/resources/IDPLicenseUsageResource.java#L94)):*
<details><summary><i>Previously, kapil-harness (Kapil Choudhary) wrote…</i></summary><blockquote>

same as above
</blockquote></details>

We accept filter / other params in these APIs. And those are send in request body and that's why it is POST. 
As per standards, we shouldn't use request body in GET request if that has a meaning on response. 
And also as per spec standards, it is not advisable to have to many parms in query / path. Instead those can converted to request body.
___
*[`idp-service/src/main/java/io/harness/idp/events/consumers/AbstractIdpServiceRedisStreamConsumer.java` line 60 at r5](https://reviewable.io/reviews/harness/harness-core/54099#-Nh1h7nrErziBCvwuM4L:-Nh5FmojD5NYSiq7W46s:btddide) ([raw file](https://github.com/harness/harness-core/blob/b6b318cedccae2e0843e969180576616757013e9/idp-service/src/main/java/io/harness/idp/events/consumers/AbstractIdpServiceRedisStreamConsumer.java#L60)):*
<details><summary><i>Previously, kapil-harness (Kapil Choudhary) wrote…</i></summary><blockquote>

.equals can handle null case, so no need for a separate null check
</blockquote></details>

Done. Used StringUtils.equals(String, String)
___
*[`idp-service/src/main/java/io/harness/idp/events/consumers/AbstractIdpServiceRedisStreamConsumer.java` line 99 at r5](https://reviewable.io/reviews/harness/harness-core/54099#-Nh1gj2zEX7Nm1mEQ8mE:-Nh5FoWw5K20EPEHVR_p:bb6un1j) ([raw file](https://github.com/harness/harness-core/blob/b6b318cedccae2e0843e969180576616757013e9/idp-service/src/main/java/io/harness/idp/events/consumers/AbstractIdpServiceRedisStreamConsumer.java#L99)):*
<details><summary><i>Previously, kapil-harness (Kapil Choudhary) wrote…</i></summary><blockquote>

this is already in finally clause, can be removed from here
</blockquote></details>

Done

**Comment by sathish-soundarapandian** [comment]
trigger uts

**Comment by sathish-soundarapandian** [comment]
trigger ut7

**Review by kapil-harness — APPROVED** [review]
> Reviewed 23 of 43 files at r1, 17 of 20 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, 4 of 4 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/54099)* status: 4 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by vikyathharekal — APPROVED** [review]
> Reviewed 4 of 4 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/54099)* status: 1. unresolved discussions.

**Comment by sathish-soundarapandian** [comment]
trigger review

**Review by mankrit-singh — APPROVED** [review]

**Review by harness-jc — APPROVED** [review]

**Pull request merged**